packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,203 @@
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateConstituents: {
+        duplicated: TSESTree.TypeNode;
+        duplicatePrevious: TSESTree.TypeNode;
+      }[] = [];
+      const cachedTypeMap: Map<TSESTree.TypeNode, Type> = new Map();
+      node.types.reduce<TSESTree.TypeNode[]>(
+        (uniqConstituents, constituentNode) => {
+          const duplicatedPreviousConstituentInAst = uniqConstituents.find(
+            ele => isSameAstNode(ele, constituentNode),
+          );
+          if (duplicatedPreviousConstituentInAst) {
+            duplicateConstituents.push({
+              duplicated: constituentNode,
+              duplicatePrevious: duplicatedPreviousConstituentInAst,
+            });
+            return uniqConstituents;
+          }
+          const constituentNodeType = checker.getTypeAtLocation(
+            parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+          );
+          const duplicatedPreviousConstituentInType = [
+            ...cachedTypeMap.entries(),
+          ].find(([, type]) => {
+            return type === constituentNodeType;
+          })?.[0];
+          if (duplicatedPreviousConstituentInType) {
+            duplicateConstituents.push({
+              duplicated: constituentNode,
+              duplicatePrevious: duplicatedPreviousConstituentInType,
+            });
+            return uniqConstituents;
+          }
+          cachedTypeMap.set(constituentNode, constituentNodeType);
+          return [...uniqConstituents, constituentNode];
+        },
+        [],
+      );
+      duplicateConstituents.forEach(duplicateConstituent => {
+        context.report({
+          data: {
+            duplicated: sourceCode.getText(duplicateConstituent.duplicated),
+            type:
+              node.type === AST_NODE_TYPES.TSIntersectionType
+                ? 'Intersection'
+                : 'Union',
+            previous: sourceCode.getText(
+              duplicateConstituent.duplicatePrevious,
+            ),
+          },
+          messageId: 'duplicate',
+          node,
+          fix: fixer => {
+            const beforeTokens = sourceCode.getTokensBefore(
+              duplicateConstituent.duplicated,
+            );
+            const afterTokens = sourceCode.getTokensAfter(
+              duplicateConstituent.duplicated,
+            );
+            const beforeUnionOrIntersectionToken = beforeTokens
+              .reverse()
+              .find(token => token.value === '|' || token.value === '&');
+            if (!beforeUnionOrIntersectionToken) {
+              return [];
+            }
+            const bracketBeforeTokens = sourceCode.getTokensBetween(
+              beforeUnionOrIntersectionToken,
+              duplicateConstituent.duplicated,
+            );
+            const bracketAfterTokens = afterTokens.slice(
+              0,
+              bracketBeforeTokens.length,
+            );
```

Comment:
```suggestion
            const bracketAfterTokens = sourceCode.getTokensAfter(
              duplicateConstituent.duplicated,
              { count: bracketBeforeTokens.length },
            );
```

Reply:
Looks very good! 
It was very helpful to learn that there are such useful options 🙏